opm/autodiff/VFPHelpers.hpp

```diff
@@ -0,0 +1,970 @@
+    }
+}
+
+
+
+
+
+
+
+/**
+ * Computes the wfr parameter according to the wfr_type_
+ * @return Production rate of oil, gas or liquid.
+ */
+template <typename T>
+static T getWFR(const T& aqua, const T& liquid, const T& vapour,
+                  const VFPProdTable::WFR_TYPE& type) {
+    switch(type) {
+        case VFPProdTable::WFR_WOR: {
+            //Water-oil ratio = water / oil
+            T wor = aqua / liquid;
+            return zeroIfNan(wor);
+        }
+        case VFPProdTable::WFR_WCT:
+            //Water cut = water / (water + oil)
+            return zeroIfNan(aqua / (aqua + liquid));
+        case VFPProdTable::WFR_WGR:
+            //Water-gas ratio = water / gas
+            return zeroIfNan(aqua / vapour);
+        case VFPProdTable::WFR_INVALID: //Intentional fall-through
+        default:
+            OPM_THROW(std::logic_error, "Invalid WFR_TYPE: '" << type << "'");
+    }
+}
+
+
+
+
+
+
+/**
+ * Computes the gfr parameter according to the gfr_type_
+ * @return Production rate of oil, gas or liquid.
+ */
+template <typename T>
+static T getGFR(const T& aqua, const T& liquid, const T& vapour,
+                  const VFPProdTable::GFR_TYPE& type) {
+    switch(type) {
+        case VFPProdTable::GFR_GOR:
+            // Gas-oil ratio = gas / oil
+            return zeroIfNan(vapour / liquid);
+        case VFPProdTable::GFR_GLR:
+            // Gas-liquid ratio = gas / (oil + water)
+            return zeroIfNan(vapour / (liquid + aqua));
+        case VFPProdTable::GFR_OGR:
+            // Oil-gas ratio = oil / gas
+            return zeroIfNan(liquid / vapour);
+        case VFPProdTable::GFR_INVALID: //Intentional fall-through
+        default:
+            OPM_THROW(std::logic_error, "Invalid GFR_TYPE: '" << type << "'");
+    }
+}
+
+
+
+
+
+
+/**
+ * Helper struct for linear interpolation
+ */
+struct InterpData {
+    InterpData() : ind_{0, 0}, inv_dist_(0.0), factor_(0.0) {}
+    int ind_[2]; //[First element greater than or equal to value, Last element smaller than or equal to value]
+    double inv_dist_; // 1 / distance between the two end points of the segment. Used to calculate derivatives and uses 1.0 / 0.0 = 0.0 as a convention
+    double factor_; // Interpolation factor
+};
+
+
+
+
+
+
+/**
+ * Helper function to find indices etc. for linear interpolation
+ */
+inline InterpData findInterpData(const double& value, const std::vector<double>& values) {
+    InterpData retval;
+
+    //If we only have one value in our vector, return that
+    if (values.size() == 1) {
+        retval.ind_[0] = 0;
+        retval.ind_[1] = 0;
+        retval.inv_dist_ = 0.0;
+        retval.factor_ = 0.0;
+    }
+    // Else search in the vector
+    else {
+        //First element greater than or equal to value
+        //Start with the second element, so that floor_iter does not go out of range
+        //Don't access out-of-range, therefore values.end()-1
```

Comment:
I had to think a lot about this, but believe it is correct. However, code that need so much thinking is better avoided, so I'd rather have .end() in the range to lower_bound and a special treatment if the end iterator is returned. That would happen when we are extrapolating (which is not treated separately from being in the last interval with the current code).